### ISSUE TITLE
[QUESTION]: How to determine the result of MPN encoding is appropriate

### ISSUE THREAD

**Issue by PussInCode** [chemprop/chemprop#312]
**What are you trying to do?**

- I tried using MPN encoding to generate molecular diagrams, which were trained according to the characteristics of the molecular diagrams.
- But I found that the model could not be trained at all, and  thought there was a problem with my mpn coding
- So I want to know how to check out which of my own MPNN encoding is appropriate

**Previous attempts**

- After research I found that  thought it was the various weights of mpn that influenced my results

- In mpn.py,the input is the same argument but the encoding result is incorrect due to the inconsistency of the weights

```
 # Input
        if self.atom_messages:
            input = self.W_i(f_atoms)  # num_atoms x hidden_size
        else:
            input = self.W_i(f_bonds)  # num_bonds x hidden_size
        message = self.act_func(input)  # num_bonds x hidden_size

        # Message passing
        for depth in range(self.depth - 1):
            if self.undirected:
                message = (message + message[b2revb]) / 2

            if self.atom_messages:
                nei_a_message = index_select_ND(message, a2a)  # num_atoms x max_num_bonds x hidden
                nei_f_bonds = index_select_ND(f_bonds, a2b)  # num_atoms x max_num_bonds x bond_fdim
                nei_message = torch.cat((nei_a_message, nei_f_bonds), dim=2)  # num_atoms x max_num_bonds x hidden + bond_fdim
                message = nei_message.sum(dim=1)  # num_atoms x hidden + bond_fdim
            else:
                # m(a1 -> a2) = [sum_{a0 \in nei(a1)} m(a0 -> a1)] - m(a2 -> a1)
                # message      a_message = sum(nei_a_message)      rev_message
                nei_a_message = index_select_ND(message, a2b)  # num_atoms x max_num_bonds x hidden
                a_message = nei_a_message.sum(dim=1)  # num_atoms x hidden
                rev_message = message[b2revb]  # num_bonds x hidden
                message = a_message[b2a] - rev_message  # num_bonds x hidden

            message = self.W_h(message)
            message = self.act_func(input + message)  # num_bonds x hidden_size
            message = self.dropout_layer(message)  # num_bonds x hidden

        a2x = a2a if self.atom_messages else a2b
        nei_a_message = index_select_ND(message, a2x)  # num_atoms x max_num_bonds x hidden
        a_message = nei_a_message.sum(dim=1)  # num_atoms x hidden
        a_input = torch.cat([f_atoms, a_message], dim=1)  # num_atoms x (atom_fdim + hidden)
        atom_hiddens = self.act_func(self.W_o(a_input))  # num_atoms x hidden
        atom_hiddens = self.dropout_layer(atom_hiddens)  # num_atoms x hidden

        # concatenate the atom descriptors
        if atom_descriptors_batch is not None:
            if len(atom_hiddens) != len(atom_descriptors_batch):
                raise ValueError(f'The number of atoms is different from the length of the extra atom features')

            atom_hiddens = torch.cat([atom_hiddens, atom_descriptors_batch], dim=1)     # num_atoms x (hidden + descriptor size)
            atom_hiddens = self.atom_descriptors_layer(atom_hiddens)                    # num_atoms x (hidden + descriptor size)
            atom_hiddens = self.dropout_layer(atom_hiddens)                             # num_atoms x (hidden + descriptor size)

        # Readout

```

- The final problem is traced due to the different initial weights of the molecules
chemprop/models/model.py，line 58
`  initialize_weights(self) `

- I don't know how to get the right weight value, is there anything else that hasn't been taken into account besides the influence of random seeds

**Screenshots**

- I'm currently using the old version, but the main problem is not getting proper mpn encoding

- old version

![image](https://user-images.githubusercontent.com/38970840/181152359-4295eb45-c2d8-4aa2-b680-e60841e2a927.png)

- new  version

mode.py line 196
![image](https://user-images.githubusercontent.com/38970840/181152093-39eb0b4d-ac1d-44bc-9e20-313e9d4b2ac9.png)

**Comment by shihchengli** [comment]
If you want to get the MPNN encoding, you can use `chemprop_fingerprint ` with your model checkpoint file(s).  More information can be found [here](https://github.com/chemprop/chemprop/tree/master#encode-fingerprint-latent-representation). If you want to freeze the weights in MPNN layers, you can use `--checkpoint_frzn ` argument with your model checkpoint file during the training of a new model.

**Comment by PussInCode** [comment]
ok， I'ii try it,thanks

**Comment by PussInCode** [comment]
Actually, I'm not in the train section right now，So I don't have checkpoint file.
At present, it is always in the mpn encode process before training.

**Comment by shihchengli** [comment]
If you want to get the MPNN encoding, you need to have a checkpoint file from a trained model. Otherwise, the MPNN encoding is not meaningful. I don't fully understand your question though. Do you just want to train a model using Chemporp but failed? Or do you want to use our MPNN model with another model architecture except for FFNN?

**Comment by PussInCode** [comment]
yes，I want to use our MPNN model with another model combined,mpnn just as a pre train

**Comment by shihchengli** [comment]
I see. Then, the `self.encoder(batch, features_batch, atom_descriptors_batch,
                                       atom_features_batch, bond_features_batch)` in line 196 in model.py is basically the MPNN encoding you are looking for, but you cannot get this without providing any checkpoint file, since it doesn't make sense to get the embedding before the model is trained. There are some weights in the MPNN model that will be learned during the backpropagation so the MPNN model and readout layers should be trained simultaneously. So far, the Chemprop is not that modularized so it is not easy to take the NPNN out and combine it with another model you want to use.  The modulization of Chemporp would happen in Chemprop 2.0. Before that happen, I would suggest you replace the `self.ffn` with the framework you want to use in line  196.

**Comment by PussInCode** [comment]
thanks,In old version,I use `self.encoder(batch, features_batch, atom_descriptors_batch, atom_features_batch, bond_features_batch)` as my encoder input feature，other model as ffn to train, but I find It's hard to train.
    I find  input feature is not right when I modified the model. I want to know the result(`self.encoder(batch, features_batch, atom_descriptors_batch, atom_features_batch, bond_features_batch)`)  comes from whether the MPN has been coded or is an intermediate state that the MPN will continue to update.

**Comment by shihchengli** [comment]
`batch, features_batch, atom_descriptors_batch, atom_features_batch, bond_features_batch` are the features of training data (SMILES, additional features, additional atom descriptors, additional atom features, and additional bond features). The `self.encoder()` function is the MPNN layers, and the weights in MPNN would be updated during the training section. Hence, you will get different output vectors with the same inputs during the training section. `initialize_weights()` will let the weights in MPNN be initialized based on the given random seed, since you already modified the MPNN model, the weights in these two model would have different initialized weights even using the same random seed.

**Comment by PussInCode** [comment]
ok，I got it,thank you very much

**Comment by PussInCode** [comment]
There are an another question,In the first epoch,the weights of the model are not updated，the chenckpoint file is also not saved. I find these two model  also have different initialized weight.

**Comment by shihchengli** [comment]
The weights will be updated for each batch in each epoch ([L116 in train/train.py](https://github.com/chemprop/chemprop/blob/master/chemprop/train/train.py#L116)). The checkpoint file will be saved even if the training hasn't started yet ([L262 in train/run_training.py](https://github.com/chemprop/chemprop/blob/master/chemprop/train/run_training.py#L262)). Could you please check if `--save_dir` be used, otherwise the model checkpoints won't be saved. Because the numbers of weights in the two models are different so the initialized weights will be different even if you use the same random seed. I'll close this issue for now, but feel free to reopen if you have more questions about the MPNN encoding. If you figure out there are some potential bugs in Chemprop, please open a new issue.

**Comment by PussInCode** [comment]
I'm sorry to bother you again！
ok，I took the `initial weights` a step further.  https://github.com/chemprop/chemprop/blob/442a1602b670f173de166f987eab64396571ee98/chemprop/nn_utils.py#L103
I carefully checked the input parameters for both models，They are almost identical.
```

---------------------------------------chemprop------------------------------
MoleculeModel(
  (encoder): MPN(
    (encoder): MPNEncoder(
      (dropout_layer): Dropout(p=0.0, inplace=False)
      (act_func): ReLU()
      (W_i_atom): Linear(in_features=133, out_features=300, bias=False)
      (W_i_bond): Linear(in_features=147, out_features=300, bias=False)
      (W_h_atom): Linear(in_features=447, out_features=300, bias=False)
      (W_h_0): Linear(in_features=300, out_features=300, bias=False)
      (W_h_1): Linear(in_features=300, out_features=300, bias=False)
      (W_o): Linear(in_features=600, out_features=300, bias=True)
      (gru): BatchGRU(
        (gru): GRU(300, 300, batch_first=True, bidirectional=True)
      )
      (lr): Linear(in_features=900, out_features=300, bias=False)
    )
  )
  (ffn): Sequential(
    (0): Dropout(p=0.0, inplace=False)
    (1): Linear(in_features=900, out_features=300, bias=True)
    (2): ReLU()
    (3): Dropout(p=0.0, inplace=False)
    (4): Linear(in_features=300, out_features=2, bias=True)
  )
)

------------------------------modified model --------------------------------
MoleculeModel(
  (encoder): MPN(
    (encoder): MPNEncoder(
      (dropout_layer): Dropout(p=0.0, inplace=False)
      (act_func): ReLU()
      (W_i_atom): Linear(in_features=133, out_features=300, bias=False)
      (W_i_bond): Linear(in_features=147, out_features=300, bias=False)
      (W_h_atom): Linear(in_features=447, out_features=300, bias=False)
      (W_h_0): Linear(in_features=300, out_features=300, bias=False)
      (W_h_1): Linear(in_features=300, out_features=300, bias=False)
      (W_o): Linear(in_features=600, out_features=300, bias=True)
      (gru): BatchGRU(
        (gru): GRU(300, 300, batch_first=True, bidirectional=True)
      )
      (lr): Linear(in_features=900, out_features=300, bias=False)
    )
  )
  (ffn): Sequential(
    (0): Dropout(p=0.0, inplace=False)
    (1): Linear(in_features=900, out_features=300, bias=True)
    (2): ReLU()
    (3): Dropout(p=0.0, inplace=False)
    (4): Linear(in_features=300, out_features=1, bias=True)
  )
)
```
The way the data is calculated is related to the mean and distribution of the data, the same data their distribution should be the same, but I get really different weight results.https://github.com/chemprop/chemprop/blob/442a1602b670f173de166f987eab64396571ee98/chemprop/nn_utils.py#L113
![image](https://user-images.githubusercontent.com/38970840/182278189-4ef30e00-d872-4d65-b782-f9b780ff39c3.png)
